### ISSUE TITLE
fetchGit and submodules

### ISSUE THREAD

**Issue by roberth** [nmattia/niv#58]
I love the support for 'normal' GitHub repo's, but `niv` in its current implementation can't include the submodules, which makes it useless in some cases.
This can be solved by adding support for `fetchGit`.

**Comment by nmattia** [comment]
Definitely. Whenever we merge #35 (_cough cough Soon ™_) I'll most likely deprecate the `url_template` fetching for github repos and use `fetchGit`/`fetchFromGitHub` instead.

**Comment by PierreR** [comment]
@nmattia I guess the resolution of this issue would solve my issue:
```
unpacking source archive /build/v3.7.2.tar.gz
error: zip file must contain a single file or directory.
hint: Pass stripRoot=false; to fetchzip to assume flat list of files.
```
I am trying to replace a call to pkgs.fetchgit with niv. Here is my source.json:
```
    "puppet-unit-test": {
        "sha256": "0hyvh2niq939rxhs7wx6szmvrs7s6nrn3sxf1rma5xl41aw3b068",
        "type": "tarball",
        "url": "http://bitbucket.cirb.lan/CICD/puppet-unit-test/archive/v3.7.2.tar.gz",
        "url_template": "http://bitbucket.cirb.lan/CICD/puppet-unit-test/archive/v3.7.2.tar.gz",
        "version": "3.7.2"
```

**Comment by nmattia** [comment]
@PierreR in this case you can also use `type = "file"` and do the unpacking yourself. But yes, having `fetchgit` and the bitbucket `git@...` uri should fix this.

**Comment by nmattia** [comment]
Forgot to update this! There's experimental support for `fetchGit`:

``` sh
$ niv add git git@github.com:nmattia/niv -a rev=deadbeef
```

**Comment by andir** [comment]
How would I use `fetchSubmodules = true` with this?

**Comment by Nadrieril** [comment]
This is not directly supported, but you can make it work as follows.
First you need to use an unstable version of nix, so that `fetchGit` accepts a `submodules` argument (it's not called `fetchSubmodules` anymore). Then change `sources.nix` to pass the `submodules` argument to `fetchGit`:
```diff
diff --git a/nix/sources.nix b/nix/sources.nix                                                                                                     
index 8a725cb..800b0f8 100644                                                                                                                      
--- a/nix/sources.nix                                                                                                                              
+++ b/nix/sources.nix                                                                                                                              
@@ -19,7 +19,7 @@ let                                                                                                                              
       pkgs.fetchzip { inherit (spec) url sha256; };                                                                                               
                                                                                                                                                   
   fetch_git = spec:                                                                                                                               
-    builtins.fetchGit { url = spec.repo; inherit (spec) rev ref; };                                                                               
+    builtins.fetchGit { url = spec.repo; inherit (spec) rev ref submodules; };
  
   fetch_builtin-tarball = spec:
     builtins.trace
```
Finally, add `"submodules": true` in a sources.json entry that has `"type": "git"`. Example:
```json
    "dhall-haskell": {
        "ref": "master",
        "repo": "https://github.com/dhall-lang/dhall-haskell",
        "rev": "9afffaea7108a86df197cb97b64a9504fe37ecfc",
        "submodules": true,
        "type": "git"
    },
```

**Comment by shajra** [comment]
@nmattia I just realized how long-standing this issue has been.  I've personally run into a second time now.  What are your latest thoughts on how best to solve it?  Is there anything I can do to help push it over the finish line?

**Comment by nmattia** [comment]
hey @shajra , I wrote up some thoughts in here: https://github.com/nmattia/niv/issues/229#issuecomment-674013591

I unfortunately don't have as much time as I used to these days, but one way to contribute would be to share the above and get feedback on it; then we can actually go forward (implementation shouldn't take long).

Also I didn't see @Nadrieril 's comment above; that's definitely something we could add to sources.nix right now! PRs welcome.
